iceberg/src/test/scala/org/apache/spark/sql/delta/him/HIM.scala

```diff
@@ -0,0 +1,167 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.him
+
+import java.io.{BufferedReader, File, InputStreamReader, IOException}
+import java.net.ServerSocket
+import java.nio.file.Files
+import java.sql.{Connection, DriverManager}
+
+import org.apache.commons.io.FileUtils
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.hive.conf.HiveConf
+import org.apache.hadoop.hive.conf.HiveConf.ConfVars
+import org.apache.hadoop.hive.metastore.HIMServer
+
+/**
+ * HIM (Hms In Memory) is an embedded Hive MetaStore for testing purpose.
+ * Multiple HIM instances can be started in parallel on the same host.
+ * see [[HIMExamples]] for how to use HIM in the code.
+ */
+class HIM {
+  private var server: HIMServer = _
+  private var whFolder: String = _
+  private var dbName: String = _
+  private var started = false
+  private var port: Int = 0
+
+  /**
+   * Start a HIM instance
+   */
+  def start(): Unit = {
+    if (started) return
+    port = HIM.firstAvailablePort()
+    val dbFolder = Files.createTempDirectory("him_metastore")
+    Files.delete(dbFolder) // Derby needs an empty folder
+    dbName = dbFolder.toString
+    whFolder = Files.createTempDirectory("him_warehouse").toString
+
+    initDatabase(dbName)
+
+    val innerConf = new HiveConf()
+    innerConf.set(ConfVars.HIVE_IN_TEST.varname, "false")
+    innerConf.set(ConfVars.METASTOREWAREHOUSE.varname, whFolder)
+    innerConf.set(ConfVars.METASTORECONNECTURLKEY.varname, s"jdbc:derby:$dbName;create=true")
+    server = new HIMServer(innerConf, port)
+    server.start()
+
+    started = true
+  }
+
+  /**
+   * Stop a HIM instance and cleanup its resources
+   */
+  def stop(): Unit = {
+    if (!started) return
+    server.stop()
+    // Cleanup on exit
+    FileUtils.deleteDirectory(new File(dbName))
+    FileUtils.deleteDirectory(new File(whFolder))
+    started = false
+  }
+
+  /**
+   * Fetch the configuration used for clients to connect to the MetaStore
+   * @return conf containing thrift uri and warehouse location
+   */
+  def conf(): Configuration = {
+    if (!started) throw new IllegalStateException("Not started")
+    val conf = new Configuration()
+    conf.set(ConfVars.METASTOREWAREHOUSE.varname, whFolder)
+    conf.set(ConfVars.METASTOREURIS.varname, s"thrift://localhost:$port")
+    conf
+  }
+
+  /**
+   * Initialize HMS schema in the Apache Derby instance.
+   * @param dbFolder the folder to create the database, also the database name
+   */
+  private def initDatabase(dbFolder: String): Unit = {
```

Comment:
Why do we need to init the DB? Isn't the HMS automatically init it during the startup? You may need to set few configs, but I remember HMS was able to init the DB during the startup.